### ISSUE TITLE
JWT is being processed when permit policy is set

### ISSUE THREAD

**Issue by Mulgikapsas** [quarkusio/quarkus#6096]
**Describe the bug**
JWT token is processed even if http policy is set to permit.

**Expected behavior**
I expect to see 200 (OK) response with metrics payload.

**Actual behavior**
I see 401 (Unauthorized) response.

**To Reproduce**
1) Define the following in microprofile-config.properties:
     ```
     quarkus.http.auth.permission.metrics.policy=permit
     quarkus.http.auth.permission.metrics.methods=GET
     quarkus.http.auth.permission.metrics.paths=/metrics
     ```

2) Execute the following:
 ` curl --request GET --url http://localhost:8080/metrics --header 'authorization: Bearer invalid' -v
`
3) Observe 401 (Unathorised response) and error in logs (Token is invalid: JWT processing failed. etc.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Windows 7
 - Output of `java -version`:  openjdk version "11.0.2" 2019-01-15
 - Quarkus version or git rev: 1.0.1.Final

**Comment by sberyozkin** [comment]
@Mulgikapsas I think we've had similar issues before, I'm sorry but I don't see it as a bug. `@PermitAll` does not mean that an invalid token can be sent along the way.
What is the use case ?

**Comment by Mulgikapsas** [comment]
@sberyozkin Prometheus is supplying JWT token when scraping /metrics page. I would like it to be ignored as there is no security on /metrics or /health endpoints by default.

**Comment by sberyozkin** [comment]
@Mulgikapsas OK, thanks. 
Here is another scenario: the user code would like to access `JsonWebToken` in the `@PermitAll` method, irrespective of what the token `groups` claim is set to. The policy configuration in your case should work on the `@PermitAll` if there is no token, but I don't think we can just dismiss the available token based on the fact it is a `permit` policy.

Can you configure Prometheus to not send the tokens or make it sending the valid tokens, why are those tokens invalid ?

Otherwise we'd need to make some code concessions to support the cases where the tokens are not meant to be sent :-). For example 1) have the in-scope policy info available to HttpAuthenticators 2) have a new policy mode, like 'ignore-credentials` 3) update all HttpAuthenticators to check the current path, and if it is `ignore-credentials` and the path matches then we skip the creds validation

This will add to the processing time and we will do it all to accommodate the tools/some client code which post the invalid credentilals :-)

Can you please check what can be done at the Prometheus level ?

CC @stuartwdouglas

**Comment by Mulgikapsas** [comment]
@sberyozkin Unfortunatey I can't change that prometheus instance. 
The token is invalid because I need to supply a different public key to SmallRye JWT for signature validation. Not sure if it is possible to specify multiple public keys to SmallRye JWT?

Another issue is that /metrics is not a JAX-RS endpoint (correct me if I am wrong), it is exposed by quarkus-metrics extension, so I am not sure how can I manipulate that endpoint to ignore Authorization header.

**Comment by stuartwdouglas** [comment]
Basically at the moment we do pro active authentication, so if there is a credential there we will attempt to verify it, and actively block invalid credentials even if the resource is unprotected. 

I think we can do this by simply having a setting that ignores auth failure and lets the request proceed. If the request does not require security then it will work as expected, if it does then the failed security check will force a challenge to be sent and valid credentials can be supplied.

**Comment by sberyozkin** [comment]
@stuartwdouglas indeed, on the secure path it would fail anyway, good point. Should we just have a new policy mode, `permit-and-ignore-authentication`, a bit long but is clear what it is ? Or something shorter may be...

@Mulgikapsas Yes, you can use JWK key set, for example, see this [test resource](https://github.com/smallrye/smallrye-jwt/blob/master/testsuite/basic/src/test/resources/publicKeySet.jwk). Make sure the token `kid` header matches one of the keys. Search for `PEM to JWK` online converter to get this set prepared quickly.

**Comment by stuartwdouglas** [comment]
In Wildfly/Undertow this is called proactive-auth, so quarkus.http.proactive-auth=false would be my preference to turn it off.

**Comment by sberyozkin** [comment]
Sounds good, we can just update the docs and say it applies to the `permit` policies

**Comment by Mulgikapsas** [comment]
Are there any plans to introduce quarkus.http.proactive-auth propery?

@sberyozkin The "kid" header is missing from the token supplied by Prometheus but is present on the other tokens.

**Comment by Mulgikapsas** [comment]
Curreny workaround is to specify Auth header filter. 
```
@ApplicationScoped
public class RequestFilters {

    /**
     * Fixes https://github.com/quarkusio/quarkus/issues/6096
     */
    public void filters(@Observes final Filters filters) {
        filters.register(rc -> {
            if ("/metrics".equals(rc.request().path())) {
                rc.request().headers().remove("Authorization");
            }
            rc.next();
        }, 9000);
    }

}
```

**Comment by devisions** [comment]
@sberyozkin @stuartwdouglas 
I am currently looking to contribute here also.

Besides adding `quarkus.http.proactive-auth` option (quite simple in `io.quarkus.vertx.http.runtime.HttpConfiguration`), looking into the following call stack I am considering altering the behaviour in `QuarkusIdentityProviderManagerImpl`.

![image](https://user-images.githubusercontent.com/14140226/77821163-afac7b80-70f0-11ea-870c-3a6d70d94aa2.png)

It's just an initial idea. Still looking into it ...

**Comment by stuartwdouglas** [comment]
So this is relatively complex, as there are a few places where the identity is accessed (e.g. the roles allowed interceptor). We may need to make some minor API changes to make sure there is a unform way to handle this instead of every consumer of the security API needing to handle it in an ad hoc manner.

**Comment by devisions** [comment]
@stuartwdouglas Thanks for the feedback, Stuart.
I gave it a first try last weekend, but implementing the aforementioned initial thought turned out to be working for _PermitAll_, but broke the _RolesAllowed_. :/
We would need this feature quite soon, so how can I help in this matter?

**Comment by stuartwdouglas** [comment]
I have made a start on this here: https://github.com/stuartwdouglas/quarkus/pull/new/proactive-authentication. 

It's quite complex, as anywhere that can potentially request the identity also has to be able to deal with failure.

**Comment by devisions** [comment]
@stuartwdouglas Hi Stuart, sorry for the late feedback, I've being caught with some work.
And many thanks for working on it! 🙏  Looking at the recent actions on this topic, it means I can play with/test it (using the master branch), right?

@sberyozkin I assume it will be included in `1.4.0.CR2`.
